src/compute/src/sink/materialized_view_v2.rs

```diff
@@ -33,5 +155,874 @@ pub(super) fn persist_sink<S>(
+                            }
+                        }
+                    }
+                    Some(event) = descs_input.next() => {
+                        match event {
+                            Event::Data(cap, data) => {
+                                for desc in data {
+                                    state.absorb_batch_description(desc, cap.clone());
+                                }
+                                state.maybe_write_batch().await
+                            }
+                            Event::Progress(_frontier) => None,
+                        }
+                    }
+                    // All inputs are exhausted, so we can shut down.
+                    else => return,
+                };
+
+                if let Some((batch, cap)) = maybe_batch {
+                    batches_output.give(&cap, batch);
+                }
+            }
+        });
+
+        let token = Box::new(button.press_on_drop());
+
+        (batches_output_stream, token)
+    }
+
+    /// State maintained by the `write` operator.
+    struct State {
+        worker_id: usize,
+        persist_writer: WriteHandle<SourceData, (), Timestamp, Diff>,
+        /// Contains `desired - persist`, reflecting the updates we would like to commit to
+        /// `persist` in order to "correct" it to track `desired`. This collection is only modified
+        /// by updates received from either the `desired` or `persist` inputs.
+        corrections: OkErr<Correction<Row>, Correction<DataflowError>>,
+        /// The frontiers of the `desired` inputs.
+        desired_frontiers: OkErr<Antichain<Timestamp>, Antichain<Timestamp>>,
+        /// The frontiers of the `persist` inputs.
+        persist_frontiers: OkErr<Antichain<Timestamp>, Antichain<Timestamp>>,
+        /// The current valid batch description and associated output capability, if any.
+        batch_description: Option<(BatchDescription, Capability<Timestamp>)>,
+    }
+
+    impl State {
+        fn new(
+            worker_id: usize,
+            persist_writer: WriteHandle<SourceData, (), Timestamp, Diff>,
+            metrics: SinkMetrics,
+        ) -> Self {
+            let worker_metrics = metrics.for_worker(worker_id);
+
+            Self {
+                worker_id,
+                persist_writer,
+                corrections: OkErr::new(
+                    Correction::new(metrics.clone(), worker_metrics.clone()),
+                    Correction::new(metrics, worker_metrics),
+                ),
+                desired_frontiers: OkErr::new_frontiers(),
+                persist_frontiers: OkErr::new_frontiers(),
+                batch_description: None,
+            }
+        }
+
+        fn advance_desired_ok_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if PartialOrder::less_than(&self.desired_frontiers.ok, &frontier) {
+                self.desired_frontiers.ok = frontier;
+            }
+        }
```

Comment:
If you want to go with `OkErr`, the `advance_*` functions could all be functions on `OkErr`, returning a bool to indicate whether they changed anything.

Reply:
I made a free-standing `fn advance(&mut Antichain, Antichain)` function instead because that also works in cases where we don't have a frontier pair (e.g. `advance_persist_frontier`).